### ISSUE TITLE
from() does not infer type correctly with TS 3.6

### ISSUE THREAD

**Issue by felixfbecker** [ReactiveX/rxjs#4992]
## Bug Report

**Current Behavior**
```
Type 'Observable<unknown>' is not assignable to type 'Observable<string>'.
  Type 'unknown' is not assignable to type 'string'.ts(2322)
```

**Reproduction**
```ts
function asObservable(input: string | ObservableInput<string>): Observable<string> {
    return typeof input === 'string' ? of(input) : from(input)
}
```

**Expected behavior**
Should compile

**Environment**
- RxJS version: 6.5.2
- TS version: 3.6.2

**Possible Solution**
Instead of using `ObservedValueOf<T>`:
```ts
export declare function from<O extends ObservableInput<any>>(input: O): Observable<ObservedValueOf<O>>
```
declare `from` like:
```ts
export declare function from<O>(input: ObservableInput<O>): Observable<O>
```
this seems to work.

**Additional context/Screenshots**
![2019-08-29 10 49 04](https://user-images.githubusercontent.com/10532611/63926033-6f4a2980-ca4b-11e9-9316-67ee60911e92.gif)

**Comment by cartant** [comment]
FWIW, I think the problem is that a `string` is also an `ObservableInput<string>`. This does not effect the same problem, AFAICT:

```ts
function asObservable(input: string | Promise<string>): Observable<string> {
  return typeof input === "string" ? of(input) : from(input);
}
```

What it is in TypeScript 3.6, IDK.

**Comment by felixfbecker** [comment]
That is the reason we wrote the type switch in that helper, but still you'd expect `ObservedValueOf<Promise<string>>` to infer `string`, since `Promise<string>` should clearly only match the `Promise<T>` in `ObservableInput<T>`

**Comment by cartant** [comment]
Hard to see this as anything other than a TypeScript bug:

```ts
declare const i: ObservableInput<string>;
const r = from(i); // Observable<unknown>
```

And one that will pretty badly wreck the typings that will be even more widely used in v7. 😭

**Comment by cartant** [comment]
@benlesh ^

**Comment by felixfbecker** [comment]
Yeah already filed in https://github.com/microsoft/TypeScript/issues/33131

But is there a need to use `ObservedValueOf` here?

**Comment by cartant** [comment]
> But is there a need to use `ObservedValueOf` here?

I think there might be, yes:

https://github.com/ReactiveX/rxjs/commit/eb1d5963178f3062aca67340532349e93232ee27

**Comment by benlesh** [comment]
Yes, specifically this doesn't work without it:

```ts
// from(Observable<A> | Observable<B>): Observable<A | B>
from(Math.random() > 0.5 ? of('123') : of(123));
```

**Comment by maxdeviant** [comment]
Not using RxJS, but I can confirm that a number of spots in our codebase where type inference was working before (in 3.4.1) broke after upgrading to 3.6.2.

**Comment by cartant** [comment]
This seems to be fixed in TypeScript 3.6.3.

**Comment by felixfbecker** [comment]
Yep, can confirm it’s fixed.
